File: test/apply_gate.py
import numpy as np


def apply_gate(psi, U, idx, nqubits):
    """
    Apply the (single or multi)-qubit gate `U` to the qubits `idx`.
    """
    assert np.size(psi) == 2**nqubits
    # convert a single integer to a tuple storing this integer
    if isinstance(idx, int): idx = (idx,)
    assert U.shape == (2**len(idx), 2**len(idx))
    if len(idx) == 1:
        # single-qubit gate
        i = idx[0]
        assert 0 <= i < nqubits
        psi = np.reshape(psi, (2**i, 2, 2**(nqubits-i-1)))
        psi = np.einsum(U, (1, 2), psi, (0, 2, 3), (0, 1, 3))
        psi = np.reshape(psi, -1)
        return psi
    elif len(idx) == 2:
        # two-qubit gate
        i, j = idx  # unpack indices
        assert 0 <= i < j < nqubits
        U = np.reshape(U, (2, 2, 2, 2))
        psi = np.reshape(psi, (2**i, 2, 2**(j-i-1), 2, 2**(nqubits-j-1)))
        psi = np.einsum(U, (1, 5, 2, 4), psi, (0, 2, 3, 4, 6), (0, 1, 3, 5, 6))
        psi = np.reshape(psi, -1)
        return psi
    elif len(idx) == 3:
        # three-qubit gate
        i, j, k = idx  # unpack indices
        assert 0 <= i < j < k < nqubits
        U = np.reshape(U, (2, 2, 2, 2, 2, 2))
        psi = np.reshape(psi, (2**i, 2, 2**(j-i-1), 2, 2**(k-j-1), 2, 2**(nqubits-k-1)))
        psi = np.einsum(U, (1, 4, 8, 2, 5, 7), psi, (0, 2, 3, 5, 6, 7, 9), (0, 1, 3, 4, 6, 8, 9))
        psi = np.reshape(psi, -1)
        return psi
    elif len(idx) == 4:
        # four-qubit gate
        i, j, k, l = idx  # unpack indices
        assert 0 <= i < j < k < l < nqubits
        U = np.reshape(U, (2, 2, 2, 2, 2, 2, 2, 2))
        psi = np.reshape(psi, (2**i, 2, 2**(j-i-1), 2, 2**(k-j-1), 2, 2**(l-k-1), 2, 2**(nqubits-l-1)))
        psi = np.einsum(U, (1, 4, 8, 11, 2, 5, 7, 10), psi, (0, 2, 3, 5, 6, 7, 9, 10, 12), (0, 1, 3, 4, 6, 8, 9, 11, 12))
        psi = np.reshape(psi, -1)
        return psi
    else:
        raise RuntimeError("currently only up to 4-qubit gates supported")

File: src/apply_gate.c
#include <assert.h>
#include "apply_gate.h"


//________________________________________________________________________________________________________________________
///
/// \brief Apply a single-qubit gate to qubit 'i' of a statevector.
///
void apply_single_qubit_gate(const struct single_qubit_gate* gate, int i, const struct statevector* restrict psi, struct statevector* restrict psi_out)
{
	assert(psi->nqubits == psi_out->nqubits);
	assert(0 <= i && i < psi->nqubits);
	const int m = 1 << i;
	const int n = 1 << (psi->nqubits - 1 - i);
	for (int k = 0; k < m; k++)
	{
		for (int j = 0; j < n; j++)
		{
			numeric x = psi->data[k*(2*n) +      j ];
			numeric y = psi->data[k*(2*n) + (n + j)];
			psi_out->data[k*(2*n) +      j ] = gate->data[0] * x + gate->data[1] * y;
			psi_out->data[k*(2*n) + (n + j)] = gate->data[2] * x + gate->data[3] * y;
		}
	}
}


//________________________________________________________________________________________________________________________
///
/// \brief Apply a two-qubit gate to qubits 'i' and 'j' of a statevector.
///
void apply_two_qubit_gate(const struct two_qubit_gate* gate, int i, int j, const struct statevector* restrict psi, struct statevector* restrict psi_out)
{
	assert(psi->nqubits == psi_out->nqubits);
	assert(0 <= i && i < psi->nqubits);
	assert(0 <= j && j < psi->nqubits);
	assert(i < j);
	const int m = 1 << i;
	const int n = 1 << (j - i - 1);
	const int o = 1 << (psi->nqubits - 1 - j);
	for (int a = 0; a < m; a++)
	{
		for (int b = 0; b < n; b++)
		{
			for (int c = 0; c < o; c++)
			{
				numeric x = psi->data[a*(4*n*o) +      b *(2*o) +      c ];
				numeric y = psi->data[a*(4*n*o) +      b *(2*o) + (o + c)];
				numeric z = psi->data[a*(4*n*o) + (n + b)*(2*o) +      c ];
				numeric w = psi->data[a*(4*n*o) + (n + b)*(2*o) + (o + c)];
				psi_out->data[a*(4*n*o) +      b *(2*o) +      c ] = gate->data[ 0] * x + gate->data[ 1] * y + gate->data[ 2] * z + gate->data[ 3] * w;
				psi_out->data[a*(4*n*o) +      b *(2*o) + (o + c)] = gate->data[ 4] * x + gate->data[ 5] * y + gate->data[ 6] * z + gate->data[ 7] * w;
				psi_out->data[a*(4*n*o) + (n + b)*(2*o) +      c ] = gate->data[ 8] * x + gate->data[ 9] * y + gate->data[10] * z + gate->data[11] * w;
				psi_out->data[a*(4*n*o) + (n + b)*(2*o) + (o + c)] = gate->data[12] * x + gate->data[13] * y + gate->data[14] * z + gate->data[15] * w;
			}
		}
	}
}
